src/Client/ClientBase.cpp

```diff
@@ -143,6 +145,78 @@ namespace ProfileEvents
 namespace
 {
 constexpr UInt64 THREAD_GROUP_ID = 0;
+
+void handleTruncateMode(DB::ASTQueryWithOutput * query_with_output, const String & out_file, String & query)
+{
+    if (!query_with_output->is_outfile_truncate)
+        return;
+
+    // Skip atomic rename for special files
+    if (out_file == "/dev/null" || out_file == "/dev/stdout" || out_file == "/dev/stderr")
+        return;
+
+    // Create a temporary file with unique suffix
+    String tmp_file = out_file + ".tmp." + DB::toString(randomSeed());
+
+    // Update the AST to use the temporary file
+    auto tmp_file_literal = std::make_shared<DB::ASTLiteral>(tmp_file);
+    query_with_output->out_file = tmp_file_literal;
+
+    // Update the query string after modifying the AST
+    query = serializeAST(*query_with_output);
+}
+
+void cleanupTempFile(const DB::ASTPtr & parsed_query)
+{
+    if (const auto * query_with_output = dynamic_cast<const DB::ASTQueryWithOutput *>(parsed_query.get()))
+    {
+        if (query_with_output->is_outfile_truncate && query_with_output->out_file)
+        {
+            const auto & tmp_file_node = query_with_output->out_file->as<DB::ASTLiteral &>();
+            String tmp_file = tmp_file_node.value.safeGet<std::string>();
+
+            // Skip cleanup for special files
+            if (tmp_file == "/dev/null" || tmp_file == "/dev/stdout" || tmp_file == "/dev/stderr")
+                return;
+
+            if (fs::exists(tmp_file))
+                fs::remove(tmp_file);
+        }
+    }
+}
+
+void performAtomicRename(const DB::ASTPtr & parsed_query)
+{
+    if (const auto * query_with_output = dynamic_cast<const DB::ASTQueryWithOutput *>(parsed_query.get()))
+    {
+        if (query_with_output->is_outfile_truncate && query_with_output->out_file)
+        {
+            const auto & tmp_file_node = query_with_output->out_file->as<DB::ASTLiteral &>();
+            String tmp_file = tmp_file_node.value.safeGet<std::string>();
+
+            // Skip rename for special files
+            if (tmp_file == "/dev/null" || tmp_file == "/dev/stdout" || tmp_file == "/dev/stderr")
+                return;
+
+            String out_file = tmp_file.substr(0, tmp_file.find(".tmp."));
```

Comment:
We should use `rfind` instead. Otherwise, `INTO OUTFILE myfile.tmp` won't work. This edge case is a good candidate for a new test condition in the stateless test, by the way
```suggestion
            String out_file = tmp_file.substr(0, tmp_file.rfind(".tmp."));
```